core/src/main/kotlin/org/evomaster/core/search/Individual.kt

```diff
@@ -68,13 +68,21 @@ abstract class Individual (trackOperator: TrackOperator? = null, traces : Mutabl
      * Returns true if the initialization actions
      * are correct (i.e. all constraints are satisfied)
      */
-    abstract fun verifyInitializationActions(): Boolean;
+    abstract fun verifyInitializationActions(): Boolean
 
     /**
      * Attempts to repair the initialization actions.
      * Initialization actions must pass the verifyInitializationAction()
      * test after this method is invoked.
      */
-    abstract fun repairInitializationActions(randomness: Randomness);
+    abstract fun repairInitializationActions(randomness: Randomness)
+
+
+    /**
+     * generate id for genes of the individual
+     */
+    open fun seeGenesIdMap() : Map<Gene, String>{
```

Comment:
this does not seem correct. A `Gene` is a mutable object. You can't have a mutable object as a key of a map, as it will leads to bug as soon as you change the state of any of the key (as those will impact the `hashCode` and `equals` methods)

Reply:
this method is used when applying archived-based mutation, return `identifier`s for all genes in this individual. 

In archived-based mutation, i need to track which gene is mutated and how it is mutated. In this case, each `Gene` requires to have an `identifier` for tracking its evolution among
`individual`. `value` of the `return` of this method is the `identifier` which is composed of gene info and rest action info.  In addition, this method is a kind of utility for getting the `identifier` for genes in this individual, and we do not use the map later. so it is fine even the state of gene changes. Moreover, there is a reason for the usage of the map type as the return:  rest action info is required to get an `identifier` for a gene, so it may take a similar search cost between getting an `identifier` and all `identifier`s. In addition, more than one genes of an individual may be mutated in evomaster. therefore, this method is developed for getting all `identifier`s. because I want to get the `identifier` by gene, I used gene as a key of the map.

Reply:
i see... but still, mutable keys are a red flag.
what about rather have the option to "bind" a Gene to an Action, and then have a `Gene.identifier` property directly  in the Gene class? (which might throw an exception if not bound to an action)
Might have some side effects on `copy()` method though.

Reply:
it other words: now an Action knows the Genes it contains, but a Gene does not know the Action it belongs to. Would it be helpful to add such info? (but it would complicate how we create and copy genes)